hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/BlockContentsValidator.java

```diff
@@ -65,50 +77,103 @@ public void validateBlocks(@NonNull final List<Block> blocks) {
+        }
+
+        int currentIndex = startIndex + 1;
+        boolean hasEventOrStateChange = false;
 
-        if (blockItems.stream().noneMatch(BlockItem::hasEventTransaction)) { // block without a user transaction
-            // A block with no user transactions contains a `block_header`, `event_headers`, `state_changes` and
-            // `state_proof`.
-            if (blockItems.stream()
-                    .skip(2) // skip block_header and round_header
-                    .limit(blockItems.size() - 3L) // skip state_proof
-                    .anyMatch(item -> !item.hasEventHeader() && !item.hasStateChanges())) {
-                Assertions.fail(
-                        "Block with no user transactions should contain items of type `block_header`, `event_headers`, `state_changes` or `state_proof`");
+        // Process items in this round until we hit the next round header or end of items
+        while (currentIndex < items.size() && !items.get(currentIndex).hasRoundHeader()) {
+            BlockItem item = items.get(currentIndex);
+
+            if (item.hasEventHeader() || item.hasStateChanges()) {
+                hasEventOrStateChange = true;
+                currentIndex++;
+            } else if (item.hasEventTransaction()) {
+                currentIndex = validateTransactionGroup(items, currentIndex);
+            } else if (item.hasTransactionResult() || item.hasTransactionOutput()) {
+                logger.error(
+                        "Found transaction result or output without preceding transaction at index {}", currentIndex);
+                Assertions.fail("Transaction result or output must be preceded by event transaction");
+            } else {
+                logger.error("Invalid item type at index {}: {}", currentIndex, item);
+                Assertions.fail("Invalid item type in round");
             }
+        }
 
-            return;
+        if (!hasEventOrStateChange) {
+            logger.error("Round starting at index {} has no event headers or state changes", startIndex);
+            Assertions.fail("Round must contain at least one event header or state change");
         }
 
-        for (int i = 0; i < blockItems.size(); i++) {
-            // An `event_transaction` SHALL be followed by a `transaction_result`.
-            if (blockItems.get(i).hasEventTransaction()
-                    && !blockItems.get(i + 1).hasTransactionResult()) {
-                Assertions.fail("Event transaction not followed by a transaction result");
-            }
+        return currentIndex;
+    }
+
+    /**
+     * Validates a transaction group (transaction + result + optional outputs).
+     * Returns the index of the next item after this group.
+     */
+    private static int validateTransactionGroup(List<BlockItem> items, int transactionIndex) {
+        if (transactionIndex + 1 >= items.size()) {
+            Assertions.fail("Event transaction at end of block with no result");
         }
```

Comment:
It seems rare it would happen, but at least until now, it has been acceptable for a block to not contain any transactions. The code in `BlockStreamManagerImpl` doesn't seem to check for the existence of any transactions before closing a block; is this assertion too strict?